README.md
# Synopsis

This project contains SKYWARN-related forms for use by SKYWARN Spotters and Net Control Operators within the NWS Wakefield, Virginia County Warning Area (CWA).  The forms can be used either with the [Winlink Express](http://www.winlink.org/tags/rms_express) amateur radio messaging software or in a standalone form using a locally-installed web server.  These forms are designed and maintained for the [Wakefield SKYWARN Amateur Radio Support Team](http://www.wx4akq.org/).

The [project homepage](http://offline.wx4akq.org) contains additional information and an introductory video with a demonstration.


# Get Notified of Updates

Announcements of updates to this project will be made over the [wx4akq-offline-announce](http://www.wx4akq.org/mailman/listinfo/wx4akq-offline-announce) e-mail list.  This is a low-traffic, announce-only mailing list anyone can subscribe to.  Please sign up for the e-mail list as part of your installation of this project.  There may be critical updates we need to let you know about!


# Multiple Tools in One package

To streamline the documentation and make it easier to understand each option, we've broken the README file down into several parts.

  * If you only want to use our custom templates for Net Control logging or Spotter reports in the Winlink Express software, see the file `README-Winlink_only.md`
  * If you are a Wakefield SKYWARN Net Control and would like the flexibility of Winlink reporting or direct upload when Internet connectivity is available, and you have a laptop or desktop PC or Mac, see the file `README-Vagrant_instance.md`
  * If you are a Wakefield SKYWARN Net Control and would like to host your own private "offline" instance on server you control, or from your laptop/desktop, see the file `README-Local_server.md`
  * If you are a Wakefield SKYWARN Net Control and would like to host a multi-user instance of the Offline Tools within an AREDN mesh or similar limited-access wireless network, see `README-Multi_user.md`


# Known Issues

For a full list of current issues or to submit a bug, refer to the [Issues](https://github.com/scrow/wx4akq-offline-tools/issues) tracker.

## Mobile Broadband Users

Users of some mobile broadband service providers (specifically AT&T Wireless, and probably some other carriers, too) may experience "Bad header line" error messages during the initial provisioning of the Vagrant virtual machine.  These errors are likely caused by the upstream proxy services utilized by the mobile broadband service.  Users encountering these error conditions will be missing some functionality or could end up with an unusable virtual machine, and will need to re-provision their virtual machine using the `utils/reprovision_machine.bat` or `utils/reprovision_machine.sh` script.  For more information, see [Issue #32](https://github.com/scrow/wx4akq-offline-tools/issues/32).


# Contributors

This project is maintained by [Steve Crow (KG4PEQ)](mailto:kg4peq@wx4akq.org) and [Reid Barden (N1VCU)](mailto:n1vcu@wx4akq.org).  Additional contributors can be found on the [Contributors](https://github.com/scrow/wx4akq-offline-tools/graphs/contributors) page.  


# License

Copyright (c) 2015-19, Steve Crow, Reid Barden.  This work is licensed under the BSD 2-clause “Simplified” License.  Please see the separate `LICENSE.md` file for further information.


CHANGELOG.md
# Change Log

## v2.0.0-beta1 Major version release - first beta

* Adds support for "dynamic content" updates - the `always_refresh` parameter in the offline file index will cause the file to be re-downloaded with each run of fo_download
* Adds multi-user support
* Adds server shutdown scripts

## v1.1.3 SSL bugfixes

* Addresses some issues relating to the SSL updates made in v1.1.2

## v1.1.2 SSL Updates

* Makes various updates throughout the application to adjust for SSL on the WX4AKQ.org domain

## v1.1.1 Minor enhancement release

* Removes the "Area Number" dropdown from the NCO form.  Effective with Ops Portal v8.11 (5/19/2016) the originating area number is no longer tracked in the logs.  This change increments the version number to `3`.
* Adds input validation to both the NCO and Spotter forms to verify input length is at least 10 characters in the location and report fields.  Server-side validation has been in place since Ops Portal v8 which throws reports into manual review if this criteria was not met.  This client side validation should reduce the number of reports requiring manual review.

## v1.1.0 Feature release

* Makes updates to the Roster Viewer function for improved compatibility with the back-end changes introduced in Ops Portal v8.4.
* Addresses Issue #44 by removing "Send to NWS and Copy WX4AKQ via Winlink" option from the NCO form.  Beginning with Ops Portal v8.0, a 20-minute digest of all reports is sent to WX4AKQ over Winlink automatically.
* Fixes Issue #40 by disabling access to the `.git/` folder by way of an `.htaccess` update.  This addresses a potential security concern for users with specific installation setups for offline functionality.
* Addresses Issue #39 by allowing NCO's to submit their own reports via the NCO form.  This brings the offline forms in line with Ops Portal v8 changes.  Self-reports do not count for NCO activity credit but are otherwise handled as any other report.
* Addresses Issue #41 by updating both the NCO and Spotter forms to support "estimated" and "unknown" times, syncing the functionality with Ops Portal v8.  This increments the version of these two forms to `2`.
* Implemented Issue #46 by adding information on the [wx4akq-offline-announce](http://www.wx4akq.org/mailman/listinfo/wx4akq-offline-announce) e-mail list in the `README.md` file.
* Moves the Change Log from `README.md` to a separate `CHANGELOG.md` file.
* Miscellaneous minor cosmetic updates.

## v1.0.5 Bugfix release

* Corrected Issue #45 which prevented the API key from being inserted into XML files when running in offline server mode.

## v1.0.4 Minor enhancement release

* Adds an `+rmsonly` build output to `mkrelease.sh`.  This build contains only the documentation, RMS Express forms, and installation batch file.

## v1.0.3 Minor enhancement release

* Adds missing license text to `.gitattributes`, `.gitignore`, and several `.htaccess` files.
* Updates the `README.md` file to provide a reference to the new project homepage URL.
* Updates the forms to point to the new project homepage URL in the footer block.

## v1.0.2 Minor enhancement release

* Changed build output to new `build/` folder
* Changed `.gitignore` to disregard `gh-pages/` folder

## v1.0.1 Bugfix release

* Corrected Issue #38 which prevented a top-level `.htaccess` file from being included in the packaged output file.

## v1.0.0 Initial general release

* Corrected Issue #36 by adding code in an attempt to avoid server timeouts during download sessions, particularly when an FCC database is downloaded and unpacked on a slow connection and slow computer.
* Minor README cleanup.

## v1.0.0-beta2 Second beta release

* Corrected Issues #27 and #31 plus assorted other documentation cleanup.
* Corrected Issue #28 which prevented proper RMS Express form operation under certain browsers.
* Corrected Issue #29 which prevented the `start_server.bat` file from working under Windows 8.
* Corrected Issue #30 which resolved a missing Vagrant boostrap file in the release distribution ZIP file.
* Addressed Issue #32 by adding a Known Issues section with content describing provisioning failures some or all mobile broadband users may encounter when spinning up a new Vagrant virtual machine over said mobile broadband connection.  Created a `utils/` folder with scripts to destroy and reprovision a bad Vagrant virtual machine.
* Corrected Issue #33 to fix a syntax error in the file copy routine in `install_rms_forms.bat`
* Corrected Issue #34 which prevented the `install_rms_forms.bat` file from working under Windows 8.
* Windows users will now have their default browser auto-launch to http://localhost:8080 after running `start_server.bat`
* Created a Change Log in the README file.

## v1.0.0-beta1 First beta release

This is the first beta release.  Please submit any issues or feature requests to the [Issues](https://github.com/scrow/wx4akq-offline-tools/issues) tracker.

## v0.1.0-alpha1 Initial alpha test

Initial alpha test release version.


index.php
<?php
/*	
#	This file is part of WX4AKQ Offline Tools
#	
#	Copyright (c) 2015-19, Steve Crow, Reid Barden
#	Licensed under the BSD 2-clause “Simplified” License
#	
#	For license information, see the LICENSE.md file or visit
#	http://wx4akq.org/software
*/

/*	Simple wrapper script to provide ability to submit to a server
	other than the one built in to RMS Express.
	
	Sample usage:  Install the .html file and index.php onto a local
	web server running httpd + PHP.  Submission of the form will
	generate a downloadable .xml file which could be saved locally
	and then attached to a Winlink message through an application
	other than RMS Express.  Additional use cases might be later
	submission via conventional e-mail.  Could also build an upload
	script which would send the files to the server using cURL or
	something similar. 
*/

// Set an execution time limit for Bug #36; will also have to perform
// some other overrides in resources/fo_download.inc.php
set_time_limit(90);

const SAVE_TO_QUEUE = 1;
const DOWNLOAD_ATTACHMENT = 2;

// functions to see if we are online
function connect_attempt($ip = '8.8.4.4', $port=53) {
	return (bool) @fsockopen($ip, $port, $errnum, $errstring, 1);
};
function isOnline() {
	// Test both Google nameservers
	if(connect_attempt('8.8.4.4',53) || connect_attempt('8.8.8.8',53)) {
		return true;
	} else {
		return false;
	};
};

// Set up things for multi-user mode
if(file_exists('mesh_mode.xml')) {
	$meshConfig = simplexml_load_file('mesh_mode.xml');
	if($meshConfig->mesh_mode_enable == 1) {
		if(isset($_SERVER['PHP_AUTH_USER'])) {
			$mesh_mode = true;
			$config_file_name = 'config.' . trim(strtolower($_SERVER['PHP_AUTH_USER'])) . '.xml';
			$mesh_operator = strtolower(trim($meshConfig->my_call));
			$is_mesh_operator = ($mesh_operator == strtolower(trim($_SERVER['PHP_AUTH_USER'])));
		} else {
			die('Multi-user mode is enabled but user authentication not detected.  Please check server configuration or see README.md for information.');
		};
	};
	// Load the mesh operator's configuration preferences, if available
	if(file_exists('config.' . trim(strtolower($meshConfig->my_call)) .'.xml')) {
		$meshOpConfig = simplexml_load_file('config.' . trim(strtolower($meshConfig->my_call)) . '.xml');
		$meshOp_always_refresh = $meshOpConfig->always_refresh;
	} else {
		$meshOp_always_refresh = 0;
	};
} else {
	$mesh_mode = false;
	$mesh_operator = null;
	$config_file_name = 'config.xml';
	$is_mesh_operator = false;
};

// Read configuration Options or redirect to configuration form
if(!file_exists($config_file_name)) {
	if(isset($_GET['form']) && ($_GET['form']!=='config') && ($_GET['form']!=='config_save')) {
		header('Location: index.php?form=config');
	};
};

$Config = array();
if(file_exists($config_file_name)) {
	$configFile = simplexml_load_file($config_file_name);
	$Config['my_call'] = trim($configFile->my_call);
	if($Config['my_call']=='') {
		if(isset($_GET['form']) && ($_GET['form']!=='config') && ($_GET['form']!=='config_save')) {
			header('Location: index.php?form=config');
		};
	};

	$Config['queue_folder'] = trim($configFile->queue_folder);
	if($Config['queue_folder']=='') {
		if($mesh_mode) {
			$Config['queue_folder'] == 'queue' . DIRECTORY_SEPARATOR . trim(strtolower($_SERVER['PHP_AUTH_USER']));
		} else {
			$Config['queue_folder']=='queue';
		};
	};
	if(!file_exists($Config['queue_folder'])) {
		mkdir($Config['queue_folder']);
	};

	$Config['api_key'] = trim($configFile->api_key);
	$Config['op_mode'] = trim($configFile->op_mode);
	if($Config['op_mode']=='') {
		$Config['op_mode'] = SAVE_TO_QUEUE;
	};
	$Config['upload_url'] = trim($configFile->upload_url);
	if($Config['upload_url'] == '') {
		$Config['upload_url'] = 'https://dev.wx4akq.org/ops/xml_upload.php';
	};
	if(trim($configFile->override_connect_detect)=='true') {
		$Config['override_connect_detect']=true;
	} else {
		$Config['override_connect_detect']=false;
	};
	if(trim($configFile->include_fcc)=='1') {
		$Config['include_fcc']=true;
	} else {
		$Config['include_fcc']=false;
	};

	if(trim($configFile->always_refresh)=='1') {
		$Config['always_refresh']=true;
	} else {
		$Config['always_refresh']=false;
	};

} else {
	if($mesh_mode) {
		$Config['my_call']=trim(strtoupper($_SERVER['PHP_AUTH_USER']));
		$Config['queue_folder'] = 'queue' . DIRECTORY_SEPARATOR . trim(strtolower($_SERVER['PHP_AUTH_USER']));
	} else {
		$Config['my_call'] = '';
		$Config['queue_folder'] = 'queue';
	};
	$Config['op_mode'] = SAVE_TO_QUEUE;
	$Config['upload_url'] = 'https://dev.wx4akq.org/ops/xml_upload.php';
	$Config['api_key'] = '';
	$Config['override_connect_detect'] = false;
	$Config['include_fcc'] = false;
	$Config['always_refresh'] = false;
	if(isset($_GET['form']) && ($_GET['form']!=='config') && ($_GET['form']!=='config_save')) {
		header('Location: index.php?form=config');
	};
};

function includeFooter() {
	echo <<< HTML
<hr/>
<p class="footer">Copyright &copy; 2015-19 Steve Crow, Reid Barden.  This work is licensed under the BSD 2-Clause &quot;Simplified&quot; License.  Update to the latest version of this form, download additional forms, request features and report bugs at the <a href="https://github.com/scrow/wx4akq-offline-tools" class="footer" TARGET="_blank">Project Homepage</a> (Internet connection required). SKYWARN&#0174 and the SKYWARN&#0174 logo are registered trademarks of the National Oceanic and Atmospheric Administration, used with permission.</p>
HTML;
}

function includeLogo() {
	echo('<img src="resources/WX4AKQ_Web_Logo.png" alt="[NWS Wakefield SKYWARN Logo]" HEIGHT=53 WIDTH=191 STYLE="vertical-align: middle; padding-right: 10px">');
}; // end function includeLogo();

switch($_SERVER['REQUEST_METHOD']) {
	case 'GET':
		if(isset($_GET['form'])) {
			$form = trim(strtolower($_GET['form']));
			switch($form) {
				case 'wx4akq_spotter_report_form':
					// Serve the WX4AKQ_Spotter_Report_Form.html file
					$fd = file_get_contents('forms/WX4AKQ_Spotter_Report_Form.html');
					$fd = str_replace('{FormServer}', $_SERVER['SERVER_NAME'], $fd);
					$fd = str_replace('{FormPort}', $_SERVER['SERVER_PORT'].dirname($_SERVER['REQUEST_URI'].'/index.php'),$fd);
					$fd = str_replace('id="appName" value="RMSExpress"', 'id="appName" value="runlocal"', $fd);
					$fd = str_replace('{Callsign}', $Config['my_call'], $fd);
					echo($fd);
					break;
				case 'wx4akq_nco_report_form':
					// Serve the WX4AKQ_NCO_Report_Form.html file
					$fd = file_get_contents('forms/WX4AKQ_NCO_Report_Form.html');
					$fd = str_replace('{FormServer}', $_SERVER['SERVER_NAME'], $fd);
					
					$uri_parts = explode('?', dirname($_SERVER['REQUEST_URI']), 2);
					$fd = str_replace('{AjaxPort}', $_SERVER['SERVER_PORT'].$uri_parts[0].'index.php',$fd);
					
					$fd = str_replace('{FormPort}', $_SERVER['SERVER_PORT'].dirname($_SERVER['REQUEST_URI'].'/index.php'),$fd);
					$fd = str_replace('id="appName" value="RMSExpress"', 'id="appName" value="runlocal"', $fd);
					$fd = str_replace('{Callsign}', $Config['my_call'], $fd);

					if(isset($_GET['edit']) && trim($_GET['edit'])!=='') {
						$edit_filename = $Config['queue_folder'].'/'.$_GET['edit'];
					};

					if(isset($_GET['edit']) && (trim($_GET['edit'])!=='') && (file_exists($edit_filename))) {
						// We are editing a report, so use some Javascript to populate the form
						$edit_js = '';
						
						$xml = new SimpleXMLElement(file_get_contents($Config['queue_folder'].'/'.$_GET['edit']));
						$edit_js .= 'document.getElementById("spotterCall").value="'.$xml->variables->spottercall.'";';
						$edit_js .= 'document.getElementById("reportInfo").value="'.$xml->variables->reportinfo.'";';
						$edit_js .= 'document.getElementById("eventLocation").value="'.$xml->variables->eventlocation.'";';

						if(trim($xml->variables->flagreport)=='yes') {
							$edit_js .= 'document.getElementById("flagReport").checked = true;';
						} else {
							$edit_js .= 'document.getElementById("flagReport").checked = false;';
						};

						$edit_js .= 'document.getElementById("reportDate").value="'.$xml->variables->reportdate.'";';
						$edit_js .= 'document.getElementById("reportTime").value="'.$xml->variables->reporttime.'";';
						$edit_js .= 'document.getElementById("eventDate").value="'.$xml->variables->eventdate.'";';
						$edit_js .= 'document.getElementById("eventTime").value="'.$xml->variables->eventtime.'";';

						if(trim($xml->variables->spotterid)=='yes') {
							$edit_js .= 'document.getElementById("spotterID").checked = true;';
						} else {
							$edit_js .= 'document.getElementById("spotterID").checked = false;';
						};

						$edit_js .= 'document.getElementById("reportComments").value="'.$xml->variables->reportcomments.'";';
						$edit_js .= 'document.getElementById("relayedVia").value="'.trim($xml->variables->relayedvia).'";';
						$edit_js .= 'document.getElementById("relayedTo").value="'.trim($xml->variables->relayedto).'";';

						if(trim($xml->variables->training)=='yes') {
							$edit_js .= 'document.getElementById("trainingYes").checked = true;';
						} else {
							$edit_js .= 'document.getElementById("trainingNo").checked = true;';
						};

						$edit_js .= 'document.getElementById("api_key").value="'.$Config['api_key'].'";';
						$edit_js .= 'document.getElementById("edit_filename").value="'.basename($edit_filename).'";';
						$edit_js .= 'document.getElementById("deleteBtn").style.display="inline";';

						if($xml->variables->formversion >= 2) {
							// Stuff specific to version 2 forms
							$edit_js .= 'document.getElementById("timeType").value="'.$xml->variables->timetype.'";';
						};

						$fd = str_replace('/* FORM_EDIT_PLACEHOLDER */',$edit_js, $fd);
					} else {
						// We are not editing, but we still need to stick an API key in here
						$edit_js = '';

						$edit_js .= 'document.getElementById("api_key").value="'.$Config['api_key'].'";';

						$fd = str_replace('/* FORM_EDIT_PLACEHOLDER */',$edit_js, $fd);
					};

					echo($fd);
					break;
				case 'upload':
					include('resources/fo_upload.inc.php');
					break;
				case 'menu':
					include('resources/fo_menu.inc.php');
					break;
				case 'config':
					include('resources/fo_config.inc.php');
					break;
				case 'config_save':
					$xml = new SimpleXMLElement('<config></config>');
					if(isset($_GET['my_call'])) {
						$xml->addChild('my_call',trim(strtoupper($_GET['my_call'])));
					} else {
						$xml->addChild('my_call',$Config['my_call']);
					};
					if(isset($_GET['op_mode'])) {
						$xml->addChild('op_mode',$_GET['op_mode']);
					} else {
						$xml->addChild('op_mode',$Config['op_mode']);
					};
					if(isset($_GET['upload_url'])) {
						$xml->addChild('upload_url',trim($_GET['upload_url']));
					} else {
						$xml->addChild('upload_url',$Config['upload_url']);
					};
					if(isset($_GET['api_key'])) {
						$xml->addChild('api_key',trim($_GET['api_key']));
					} else {
						$xml->addChild('api_key',$Config['api_key']);
					};
					if(isset($_GET['queue_folder']) && (trim($_GET['queue_folder']!==''))) {
						$xml->addChild('queue_folder',trim($_GET['queue_folder']));
					} else {
						$xml->addChild('queue_folder',trim($_GET['queue_folder']));
					};
					if(isset($_GET['include_fcc']) && (trim($_GET['include_fcc']!==''))) {
						$xml->addChild('include_fcc',1);
					} else {
						$xml->addChild('include_fcc',0);
					};
					if(isset($_GET['always_refresh']) && (trim($_GET['always_refresh']!==''))) {
						$xml->addChild('always_refresh',1);
					} else {
						$xml->addChild('always_refresh',0);
					};
					if(isset($_GET['override_connect_detect']) && (trim($_GET['override_connect_detect']=='true'))) {
						$xml->addChild('override_connect_detect','true');
					} else {
						$xml->addChild('override_connect_detect','false');
					};
					$fp = fopen($config_file_name,'w');
					fwrite($fp, $xml->asXML());
					fclose($fp);
					header('Location: index.php?form=menu');
					break;
				case 'download':
					include('resources/fo_download.inc.php');
					break;
				case 'editreport':
					include('resources/fo_editreport.inc.php');
					break;
				case 'roster':
					include('resources/fo_roster.inc.php');
					break;
				case 'ajax':
					error_log('trigger');
					if(isset($_GET['req'])) {
						switch($_GET['req']) {
							case 'fcc':
								// Do FCC database dip and return string formatted for NCO Form
								if(file_exists('data/fcc.sqlite3')) {
									$db = new PDO('sqlite:data/fcc.sqlite3');
									
									$found = false;
									
									// Try all three tables, smallest to largest
									
									$sth = $db->prepare('SELECT * FROM fcc_express WHERE callsign=?');
									$sth->execute(array(strtoupper(trim($_GET['callsign']))));
									$result = $sth->fetch(PDO::FETCH_ASSOC);
									if($result['callsign'] == strtoupper(trim($_GET['callsign']))) {
										$found = true;
									};
									
									if(!$found) {
										$sth = $db->prepare('SELECT * FROM fcc_regional WHERE callsign=?');
										$sth->execute(array(strtoupper(trim($_GET['callsign']))));
										$result = $sth->fetch(PDO::FETCH_ASSOC);
										if($result['callsign'] == strtoupper(trim($_GET['callsign']))) {
											$found = true;
										};
									};

									if(!$found) {
										$sth = $db->prepare('SELECT * FROM fcc_remainder WHERE callsign=?');
										$sth->execute(array(strtoupper(trim($_GET['callsign']))));
										$result = $sth->fetch(PDO::FETCH_ASSOC);
										if($result['callsign'] == strtoupper(trim($_GET['callsign']))) {
											$found = true;
										};
									};
									
									if(($found) && (trim($result['last'])!=='')) {
										if(trim($result['address']=='')) {
											$address = '';
											$divHeight = '42px';
											$insertLink = '';
										} else {
											$address = $result['address'].'<br/>';
											$divHeight = '82px';
											$insertLink = "<br/><A HREF=\"#\" onClick=\"javascript:document.getElementById('eventLocation').value='".$result['address'].", ".$result['city']." ".$result['state']."';document.getElementById('reportInfo').select();\">Use as Event Location</A>";
										};
										echo('<LABEL FOR="fccdata" style="height: '.$divHeight.'">FCC Data:</LABEL><div style="height:'.$divHeight.'">'.$result['last'].', '.$result['first'].' '.$result['middle'].'<br>'.$address.$result['city'].' '.$result['state'].' '.$result['zip'].$insertLink.'</div>');
									} else {
										echo('<LABEL FOR="fccdata">FCC Data:</LABEL>Call sign not found.');
										// return nothing, callsign not found
									};
									unset($db);
								} else {
									// return nothing, database does not exist
								};
								break;
							default:
								die('Unrecognized AJAX request.');
								break;
						}; // end switch($POST['req']);
					} else {
						die('AJAX request not supplied.');
					};
					break;
				default:
					die('Unrecognized form name.');
					break;
			}; // end switch($form);
		} else {
			// No form name specified
			// Eventually we could show a list of forms or some other default page here
			// For now, let's redirect to the Spotter form
			header('Location: index.php?form=menu');
		};
		break;
	case 'POST':
		// Handle form submissions
		// Convert array keys to lowercase for consistency with RMS Express
		$_POST = array_change_key_case($_POST, CASE_LOWER);
		$xml = new SimpleXMLElement('<RMS_Express_Form></RMS_Express_Form>');
		$xmlVariables = $xml->addChild('variables');
		foreach(array_keys($_POST) as $thiskey) {
			$xmlVariables->addChild($thiskey, $_POST[$thiskey]);
		};
		// Generate a filename
		if(isset($_POST['edit_filename']) && (trim($_POST['edit_filename'])!=='')) {
			$filename = trim($_POST['edit_filename']);
		} else {
			$filename = $_POST['formname'].'-'.time().'.xml';
		};
		
		// Delete the report if so requested
		if(isset($_POST['formname']) && ($_POST['formname']=='WX4AKQ_NCO_Report_Form') && ($_POST['operation']=='delete')) {
			if(file_exists($Config['queue_folder'].'/'.$filename)) {
				unlink($Config['queue_folder'].'/'.$filename);
			};
			die();
		};
		
		// Send the file to the browser (we might also want to provide a config option to save to disk)
		switch($Config['op_mode']) {
			case SAVE_TO_QUEUE:
				$fp = fopen($Config['queue_folder'].'/'.$filename, 'w');
				fwrite($fp, $xml->asXML());
				fclose($fp);
				break;
			case DOWNLOAD_ATTACHMENT:
				header('Content-type: text/xml');
				header('Content-Disposition: attachment; filename="'.$filename.'"');
				echo $xml->asXML();
				break;
		}; // end switch($Config['op_mode']);
		break;
	default:
		die('Unsupported request method.');
		break;
} // end switch($_SERVER['REQUEST_METHOD']);
?>
